popup_offer/popup_offer.module

```diff
@@ -1,38 +0,0 @@
-<?php
-
-/**
- * Implements hook_preprocess_html().
- */
-function popup_offer_preprocess_html(&$variables)
-{
-  /* @var \Drupal\user\UserDataInterface $userData */
-  $userData = \Drupal::service('user.data');
-  $user = \Drupal::currentUser();
-  $product_builder = \Drupal::service('commerce_product.variation_field_renderer');
-  $builder = \Drupal::service('poppup_offer.special_discount');
-  $p_datas = $userData->get('popup_offer', $user->id(), 'discount_products');
-  if (!empty($p_datas)) {
```

Comment:
also it seems you're rendering popup every time page loads. I don't think it's correct. It should be only on adding to cart product which is specified in promotion.

Reply:
I agree, just in form submit and on closing button I was need to add removing data to prevent showing popup again.